cloudflare/data_source_waf_rules.go

```diff
@@ -0,0 +1,204 @@
+						},
+						"group_id": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"rules": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"description": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"priority": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"mode": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"group_id": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"group_name": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"package_id": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceCloudflareWAFRulesRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	zoneID := d.Get("zone_id").(string)
+
+	// Prepare the filters to be applied to the search
+	filter, err := expandFilterWAFRules(d.Get("filter"))
+	if err != nil {
+		return err
+	}
+
+	// If no package ID is given, we will consider all for the zone
+	packageID := d.Get("package_id").(string)
+	var pkgList []cloudflare.WAFPackage
+	if packageID == "" {
+		var err error
+		log.Printf("[DEBUG] Reading WAF Packages")
+		pkgList, err = client.ListWAFPackages(zoneID)
+		if err != nil {
+			return err
+		}
+	} else {
+		pkgList = append(pkgList, cloudflare.WAFPackage{ID: packageID})
+	}
+
+	log.Printf("[DEBUG] Reading WAF Rules")
+	ruleDetails := make([]interface{}, 0)
+	for _, pkg := range pkgList {
+		ruleList, err := client.ListWAFRules(zoneID, pkg.ID)
+		if err != nil {
+			return err
+		}
+
+		foundGroup := false
+		for _, rule := range ruleList {
+			if filter.GroupID != "" {
+				if filter.GroupID != rule.Group.ID {
+					continue
+				}
+
+				// Allows to stop querying the API faster
+				foundGroup = true
+			}
+
+			if filter.Description != nil && !filter.Description.Match([]byte(rule.Description)) {
+				continue
+			}
+
+			if filter.Mode != "" && filter.Mode != rule.Mode {
+				continue
+			}
+
+			ruleDetails = append(ruleDetails, map[string]interface{}{
+				"id":          rule.ID,
+				"description": rule.Description,
+				"priority":    rule.Priority,
+				"mode":        rule.Mode,
+				"group_id":    rule.Group.ID,
+				"group_name":  rule.Group.Name,
+				"package_id":  pkg.ID,
+			})
+		}
+
+		if foundGroup {
+			// We can stop looking further as a group is only part of a unique
+			// package, meaning that if we found the group, no need to go look
+			// at other packages
+			break
+		}
+	}
+
+	err = d.Set("rules", ruleDetails)
+	if err != nil {
+		return fmt.Errorf("Error setting WAF rules: %s", err)
+	}
+
+	d.SetId("WAFRules " + time.Now().UTC().String())
```

Comment:
Using this as the non-unique ID here isn't my ideal as it doesn't match the other formats but we can move forward with it providing that we swap the space for an underscore to make it possible to lookup without any spaces.
```suggestion
	d.SetId(fmt.Sprintf("WAFRules_%s", time.Now().UTC().String()))
```